src/external/wannierio.jl

```diff
@@ -0,0 +1,693 @@
+            end
+        end
+        push!(win, :kpoint_path => kpoint_path)
+    end
+
+    # Return a NamedTuple so that the user can access the parameters using
+    # win.num_wann, win.num_bands, etc. (with Dict it is win[:num_wann])
+    NamedTuple(win)
+end
+
+"""
+Preprocess the scf results for Wannierization.
+
+Unfold symmetries, exclude bands.
+
+# Arguments
+- `exclude_bands`: the band indices to be excluded from the Wannierization,
+    usually the semicore states.
+"""
+function unfold_scfres_wannier(
+    scfres::NamedTuple,
+    exclude_bands::Union{AbstractArray{<:Integer},Nothing}=nothing,
+)
+    @assert scfres.basis.model.spin_polarization in (:none, :spinless, :collinear)
+
+    # Undo symmetry operations to get full k-point list
+    # for collinear calculation, I assume there is no symmetry
+    if scfres.basis.model.spin_polarization == :collinear
+        @assert scfres.basis.symmetry == :none  # TODO check
+        scfres_unfold = scfres
+    else
+        scfres_unfold = unfold_bz(scfres)
+    end
+    basis = scfres_unfold.basis
+    ψ = scfres_unfold.ψ
+    eigenvalues = scfres_unfold.eigenvalues
+
+    if !isnothing(exclude_bands)
+        n_bands = length(eigenvalues[1])
+        keep_bands = setdiff(1:n_bands, exclude_bands)
+        ψ = [ψk[:, keep_bands] for ψk in ψ]
+        eigenvalues = [e[keep_bands] for e in eigenvalues]
+    end
+
+    (; basis, ψ, eigenvalues)
+end
+
+"""
+Compute & save matrices for Wannierization.
+
+# Arguments
+- `basis`: `PlaneWaveBasis`
+- `ψ`: the wavefunctions
+- `eigenvalues`: the eigenvalues
+    usually obtained from
+    `basis, ψ, eigenvalues = _unfold_scfres(scfres, exclude_bands)`
+
+# Keyword arguments
+- `fileprefix`: the filename prefix for saving the matrices, e.g., if
+    `fileprefix = "wannier/silicon"`, `amn` will be written to
+    `wannier/silicon.amn`.
+- `n_wann`: number of Wannier functions
+- `nnkp`: a tuple of `(kpb_k, kpb_G)` specifying the ``b``-vectors.
+    Can be generated either by `Wannier90` or `Wannier.jl`.
+- `spin`: spin channel. In `:collinear` case, `1` for spin-up, `2` for spin-down.
+- `M`: the overlap matrix computed by [`compute_mmn`](@ref).
+    If given, the `compute_mmn` will be skipped.
+- `A`: the initial guess for the Wannier functions, computed by
+    [`compute_amn`](@ref).
+    If given, the `compute_amn` will be skipped.
+- `E`: the eigenvalues on a uniform grid, computed by [`compute_eig`](@ref).
+    If given, the `compute_eig` will be skipped.
+- `unk`: whether to compute real-space wavefunction files `UNK`.
+    Warning: this takes a lot of memory and time.
+- remaining keyword arguments are passed to `get_wannier90_win` and written
+    to the `win` file.
+
+!!! warning
+
+    For simplicity, here only use random initial guess for Wannier functions.
+    In principle, one should use better initial guesses, e.g.,
+    [`guess_amn_hydrogenic`](@ref), [`guess_amn_psp`](@ref).
+"""
+function save_wannier(
```

Comment:
Is there a way to run Wannier.jl without saving these?

Reply:
Also I don't like this monolithic function. I think it's much better to have individual functions to write these files (e.g. by providing DFTK-specific methods to `WannierIO.write_amn` etc.) and then the user just has to call them in the right order and supplying the right guess methods etc.